Core/MIPS/IR/IRPassSimplify.cpp

```diff
@@ -744,3 +745,364 @@ bool ReduceLoads(const IRWriter &in, IRWriter &out) {
+		case IROp::Madd:
+		case IROp::MaddU:
+		case IROp::Msub:
+		case IROp::MsubU:
+		case IROp::Div:
+		case IROp::DivU:
+			if (inst.src1 && otherRegs[inst.src1] != RegState::CHANGED)
+				otherRegs[inst.src1] = RegState::READ;
+			if (inst.src2 && otherRegs[inst.src2] != RegState::CHANGED)
+				otherRegs[inst.src2] = RegState::READ;
+			otherQueue.push_back(inst);
+			queuing = true;
+			break;
+
+		case IROp::MfHi:
+		case IROp::MfLo:
+		case IROp::FpCondToReg:
+			otherRegs[inst.dest] = RegState::CHANGED;
+			otherQueue.push_back(inst);
+			queuing = true;
+			break;
+
+		case IROp::MtHi:
+		case IROp::MtLo:
+			if (inst.src1 && otherRegs[inst.src1] != RegState::CHANGED)
+				otherRegs[inst.src1] = RegState::READ;
+			otherQueue.push_back(inst);
+			queuing = true;
+			break;
+
+		case IROp::Nop:
+		case IROp::Downcount:
+		case IROp::ZeroFpCond:
+			if (queuing) {
+				// These are freebies.  Sometimes helps with delay slots.
+				otherQueue.push_back(inst);
+			} else {
+				out.Write(inst);
+			}
+			break;
+
+		default:
+			flushQueue();
+			out.Write(inst);
+			break;
+		}
+	}
+
+	// Can reuse the old constants array - not touching constants in this pass.
+	for (u32 value : in.GetConstants()) {
+		out.AddConstant(value);
+	}
+	return logBlocks;
+}
+
+bool MergeLoadStore(const IRWriter &in, IRWriter &out) {
+	bool logBlocks = false;
+
+	auto opsCompatible = [&](const IRInst &a, const IRInst &b, int dist) {
+		if (a.op != b.op || a.src1 != b.src1) {
+			// Not similar enough at all.
+			return false;
+		}
+		u32 off1 = in.GetConstants()[a.src2];
+		u32 off2 = in.GetConstants()[b.src2];
+		if (off1 + dist != off2) {
+			// Not immediately sequential.
+			return false;
+		}
+
+		return true;
+	};
+
+	for (int i = 0, n = (int)in.GetInstructions().size(); i < n; i++) {
+		IRInst inst = in.GetInstructions()[i];
+		int c = 0;
+		switch (inst.op) {
+		case IROp::Store8:
+			for (c = 1; c < 4 && i + c < n; ++c) {
+				const IRInst &nextInst = in.GetInstructions()[i + c];
+				// TODO: Might be nice to check if this is an obvious constant.
+				if (inst.src3 != nextInst.src3 || inst.src3 != 0) {
+					break;
+				}
+				if (!opsCompatible(inst, nextInst, c)) {
+					break;
+				}
+			}
+			// Warning: this may generate unaligned stores.
+			if (c == 2 || c == 3) {
+				inst.op = IROp::Store16;
+				out.Write(inst);
+				// Skip the next one.
+				++i;
+				continue;
+			}
+			if (c == 4) {
+				inst.op = IROp::Store32;
+				out.Write(inst);
+				// Skip all 4.
+				i += 3;
+				continue;
+			}
+			out.Write(inst);
+			break;
+
+		case IROp::Store16:
+			for (c = 1; c < 2 && i + c < n; ++c) {
+				const IRInst &nextInst = in.GetInstructions()[i + c];
+				// TODO: Might be nice to check if this is an obvious constant.
+				if (inst.src3 != nextInst.src3 || inst.src3 != 0) {
+					break;
+				}
+				if (!opsCompatible(inst, nextInst, c * 2)) {
+					break;
+				}
+			}
+			// Warning: this may generate unaligned stores.
+			if (c == 2) {
```

Comment:
This might also be a good place to apply the lb -> lbu optimization and such things.

I was thinking, maybe there's some value in caching the post-optimization IR (e.g. with a hash), just like with shaders.

-[Unknown]